modules/images/webp-uploads/load.php

```diff
@@ -157,6 +166,56 @@ function webp_uploads_create_sources_property( array $metadata, $attachment_id )
 
 add_filter( 'wp_generate_attachment_metadata', 'webp_uploads_create_sources_property', 10, 2 );
 
+/**
+ * Filter the image editor default output format mapping to select the most appropriate
+ * output format depending on desired output formats and supported mime types by the image
+ * editor.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $output_format The image editor default output format mapping.
+ * @param string $filename      Path to the image.
+ * @param string $mime_type     The source image mime type.
+ * @return string The new output format mapping.
+ */
+function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
+	// Skip conversion when creating the `-scaled` image (for large image uploads).
+	if ( preg_match( '/-scaled\..{3}.?$/', $filename ) ) {
+		return $output_format;
+	}
+
+	// Use the original mime type if this type is allowed.
+	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
+	if (
+		! isset( $valid_mime_transforms[ $mime_type ] ) ||
+		! is_array( $valid_mime_transforms[ $mime_type ] ) ||
+		in_array( $mime_type, $valid_mime_transforms[ $mime_type ], true )
+	) {
+		return $output_format;
+	}
+
+	$new_mime_type = null;
+
+	// Find the first supported mime type by the image editor to use it as the default one.
+	foreach ( $valid_mime_transforms[ $mime_type ] as $target_mime ) {
+		if ( wp_image_editor_supports( array( 'mime_type' => $target_mime ) ) ) {
```

Comment:
This condition does not seem required as the editor will check if the mime is supported. I agree this helps avoid passing incorrect values but that also won't work in creating the `sources` attributes anyways so the condition can be ignored.

Reply:
This condition is required to make sure that we don't use image formats that are not supported by the server.

Reply:
Right, however this is already checked by the editor image right after this is called:

- https://github.com/WordPress/wordpress-develop/blob/f0dfa682a5cf2004f468c18ba8d5c87c463cdeea/src/wp-includes/class-wp-image-editor.php#L370

Reply:
And? Not sure what you suggest... Not checking whether the mime type is supported and blindly return the first one? My understanding is that we need to return the first **supported** mime type to make sure everything works correctly and the image editor doesn't need to fallback to the [default format](https://github.com/WordPress/wordpress-develop/blob/f0dfa682a5cf2004f468c18ba8d5c87c463cdeea/src/wp-includes/class-wp-image-editor.php#L376-L378).

Reply:
Sorry for the confusion, my suggestion here is around the need for the conditional as if you provide a nonsupported mime type from the filter the editor won't work anyway due which I think should be the expected behavior instead of changing let's say `image/avif` to `image/jpeg` if `image/avif` is not supported no image should be generated instead as currently, it would create images in a non requested mime type.

Reply:
I agree with @eugene-manuilov that this check is needed to avoid trying to create a MIME type that isn't supported and therefore will fail. This check was also part of the original module implementation from #4 and should be kept for the same purpose.